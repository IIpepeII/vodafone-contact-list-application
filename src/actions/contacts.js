/*
* Action creators for contacts.
*/

// Handle changes in contactsAreLoading state
export const contactsAreLoading = bool => ({
  type: 'CONTACTS_ARE_LOADING',
  isLoading: bool,
});

// Save initial contacts in contacts state
export const setContacts = contacts => ({
  type: 'SET_CONTACTS',
  contacts,
});

/*
* Fetch contacts asynchronously from the https://randomuser.me API.
*/
export const startSetContacts = () => async (dispatch) => {
  // Set contactsAreLoading state to true.
  dispatch(contactsAreLoading(true));
  const url = 'https://randomuser.me/api/?results=9&nat=us';

  // Every data in the API is lowercase. It is a function to capitalize strings if necessery.
  const capitalize = str => str.charAt(0).toUpperCase() + str.slice(1);

  try {
    // Fetch data
    const response = await fetch(url);
    const data = await response.json();
    const results = await data.results;

    // Create contacts array
    const users = results.map(user => ({
      id: user.id.value,
      firstName: capitalize(user.name.first),
      lastName: capitalize(user.name.last),
      email: user.email,
      cell: user.cell,
    }));

    // Set contactsAreLoading state back to false.
    dispatch(contactsAreLoading(false));

    // Dispatch contacts to Redux store
    dispatch(setContacts(users));
  } catch (e) {
    console.error(e);
  }
};

// Handle delete contact. It is done by id.
export const deleteContact = id => ({
  type: 'DELETE_CONTACT',
  id,
});

// Handle update contact. It is done by id.
export const editContact = (id, updatedData) => ({
  type: 'EDIT_CONTACT',
  id,
  updatedData,
});

// Handle add new contact to the contacts state
export const addContact = contact => ({
  type: 'ADD_CONTACT',
  contact,
});
