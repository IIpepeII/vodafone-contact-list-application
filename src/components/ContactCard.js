/*
* Contact item for Dashboard contacts grid.
* props:
* - data: individual contact object from contacts state getting in Dashboard component.
*/
import React from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';

const ContactCard = (props) => {
  const {
    id,
    firstName,
    lastName,
    email,
    cell,
  } = props.data;

  return (
    <div className="contact-card col-sm-6 col-md-4 text-center">
      <Link to={`/contact/${id}`} className="contact-card__clickwrapper">
        <header className="contact-card__header py-3">
          { lastName }, { firstName }
        </header>
        <div className="contact-card__content">
          <small className="pt-2">{ email }</small>
          <small className="py-2">{ cell }</small>
        </div>
      </Link>
    </div>
  );
};

ContactCard.propTypes = {
  data: PropTypes.objectOf(PropTypes.string).isRequired,
};

export default ContactCard;
