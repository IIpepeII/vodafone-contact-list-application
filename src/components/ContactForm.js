/*
* Form for adding and editing contact data.
* Used in ./AddContactPage.js and EditContactPage.js
* props:
* - contact: actual contact data if using the form for editing contact in <EditContactPage />.
* - page: needed to properly render text on form button. 'Save contact' or 'Add contact'.
* - getFormData: function for sending data to parent. <AddContactPage /> or <EditContactPage />
*/
import React, { Component } from 'react';
import {
  Button,
  Form,
  FormGroup,
  Label,
  Input,
} from 'reactstrap';
import PropTypes from 'prop-types';

class ContactForm extends Component {
  static propTypes = {
    contact: PropTypes.objectOf(PropTypes.string),
    page: PropTypes.string.isRequired,
    getFormData: PropTypes.func.isRequired,
  }

  // If form is child of <AddContactPage /> there is no contact prop.
  static defaultProps = {
    contact: {
      firstName: '',
      lastName: '',
      email: '',
      cell: '',
    },
  };

  state = {
    contact: {
      firstName: this.props.contact.firstName,
      lastName: this.props.contact.lastName,
      email: this.props.contact.email,
      cell: this.props.contact.cell,
    },
    error: false,
  }

  // By onchanging form fields update local state by field name property.
  updateContactState = (e) => {
    const { name, value } = e.target;
    this.setState(prevState => ({
      contact: {
        ...prevState.contact,
        [name]: value,
      },
    }));
  }

  // Handle submit form
  handleClickSaveContact = (e) => {
    e.preventDefault();

    const {
      firstName,
      lastName,
      email,
      cell,
    } = this.state.contact;

    /*
    * Regex for email validation.
    */
    const emailReg = /^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/;

    // Form validation. Check if any field is empty and email address is valid based on regex
    if (firstName.trim().length === 0
      || lastName.trim().length === 0
      || email.trim().length === 0
      || cell.trim().length === 0) {
      this.setState({ error: 'Please, fill out every field!' });
    } else if (!emailReg.test(email)) {
      this.setState({ error: 'Please, enter a valid email address!' });
    } else {
      this.props.getFormData(this.state.contact);
    }
  }

  render() {
    return (
      <Form onSubmit={this.handleClickSaveContact}>
        <FormGroup>
          <Label for="firstName">First name:</Label>
          <Input
            type="text"
            value={this.state.contact.firstName}
            name="firstName"
            placeholder="Enter First Name"
            onChange={this.updateContactState}
          />
        </FormGroup>
        <FormGroup>
          <Label for="lastName">Last name:</Label>
          <Input
            type="text"
            value={this.state.contact.lastName}
            name="lastName"
            placeholder="Enter Last Name"
            onChange={this.updateContactState}
          />
        </FormGroup>
        <FormGroup>
          <Label for="lastName">E-mail:</Label>
          <Input
            type="text"
            value={this.state.contact.email}
            name="email"
            placeholder="Enter Email"
            onChange={this.updateContactState}
          />
        </FormGroup>
        <FormGroup>
          <Label for="cell">Phone:</Label>
          <Input
            type="text"
            value={this.state.contact.cell}
            name="cell"
            placeholder="Enter Phone"
            onChange={this.updateContactState}
          />
        </FormGroup>
        {
          this.state.error && <p className="error">{this.state.error}</p>
        }
        <div className="text-center pt-4">
          <Button className="btn btn--empty btn--gray-empty">
            {this.props.page === 'editContact' ? 'Save Contact' : 'Add Contact'}
          </Button>
        </div>
      </Form>
    );
  }
}

export default ContactForm;
