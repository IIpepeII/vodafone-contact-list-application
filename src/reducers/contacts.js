/*
* Handle actions related to contacts.
*/

/*
* While contacts are fetching from the API, I set this state to true in order to display loader.
* Default state is false.
*/
export const contactsAreLoading = (state = false, action) => {
  switch (action.type) {
    case 'CONTACTS_ARE_LOADING':
      return action.isLoading;
    default:
      return state;
  }
};

/*
* Handle actions related to the contacts state.
* Default state is an empty array.
*/
export const contacts = (state = [], action) => {
  switch (action.type) {
    case 'SET_CONTACTS':
      return action.contacts;
    case 'ADD_CONTACT':
      return [
        action.contact,
        ...state,
      ];
    case 'EDIT_CONTACT':
      return state.map((contact) => {
        if (contact.id === action.id) {
          return {
            ...contact,
            ...action.updatedData,
          };
        }
        return contact;
      });
    case 'DELETE_CONTACT':
      return state.filter(contact => contact.id !== action.id);
    default:
      return state;
  }
};
