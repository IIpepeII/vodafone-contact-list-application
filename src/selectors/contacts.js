/*
* Filter function for testing matches in the last name. Used in components/Dashboard.js
*/
export default (contacts, text) => contacts.filter((contact) => {
  const regex = new RegExp(`^${text}`, 'i');
  return contact.lastName.match(regex);
});
