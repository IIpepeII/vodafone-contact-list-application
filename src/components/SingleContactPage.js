/*
* Single Contact Page
* Props:
* - contact: actual contact data from contacts state by id in URL.
* - dispatchDeleteContact for sending contact id to store via deleteContact action creator to delete the contact
* - React Router history for dropping user back to dashboard after deleting the contact
* Bootstrap 4 modal is used to confirm delete intention.
*/
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import {
  Button,
  Modal,
  ModalHeader,
  ModalBody,
  ModalFooter,
} from 'reactstrap';
import PropTypes from 'prop-types';
import { deleteContact } from '../actions/contacts';

export class SingleContactPage extends Component {
  static propTypes = {
    contact: PropTypes.objectOf(PropTypes.string),
    dispatchDeleteContact: PropTypes.func.isRequired,
    history: PropTypes.object.isRequired,
  }

  static defaultProps = {
    contact: {},
  }

  // By clicking delete button change state to true in order to open modal
  state = {
    modal: false,
  }

  // Changing modal state to open and close modal
  toggleModal = () => {
    this.setState(prevState => ({ modal: !prevState.modal }));
  }

  render() {
    const {
      id,
      firstName,
      lastName,
      email,
      cell,
    } = this.props.contact;

    return (
      <div className="single-contact">
        <h1>{firstName} {lastName}&apos;s Profile</h1>
        <div className="single-contact__data px-lg-7 pb-5">
          <p><span className="label-color">Name:</span> {firstName} {lastName}</p>
          <p><span className="label-color">E-mail:</span> <a href={`mailto:${email}`}>{email}</a></p>
          <p><span className="label-color">Phone:</span> {cell}</p>
        </div>
        <div className="text-center">
          {/* By clicking delete button we open modal to confirm */}
          <button
            type="button"
            className="btn btn--empty btn--red-empty mr-4"
            onClick={this.toggleModal}
          >
            Delete
          </button>
          <Link
            className="btn btn--empty btn--gray-empty"
            to={`/editcontact/${id}`}
          >
            Edit
          </Link>
        </div>
        {/* Bootstrap 4 modal */}
        <Modal isOpen={this.state.modal} toggle={this.toggleModal} className="">
          <ModalHeader toggle={this.toggleModal}>Delete contact</ModalHeader>
          <ModalBody>
            Are you sure you want to delete {firstName} {lastName}&apos;s Profile?
          </ModalBody>
          <ModalFooter>
            {/* By clicking delete button in modal we actually delete the contact */}
            <Button
              className="btn btn--empty btn--red-empty"
              id="delete"
              onClick={() => {
                this.props.dispatchDeleteContact(this.props.contact.id);
                this.props.history.push('/');
              }}
            >
              Delete
            </Button>{' '}
            <Button
              className="btn btn--empty btn--gray-empty"
              onClick={this.toggleModal}
            >
              Cancel
            </Button>
          </ModalFooter>
        </Modal>
      </div>
    );
  }
}

/*
* Based on React Router's match object find contact in contacts state array by id.
*/
const mapStateToProps = (state, props) => ({
  contact: state.contacts.find(contact => contact.id === props.match.params.id),
});

/*
* Create dispatchDeleteContact function to link deleteContact action creator with the component via props.
*/
const mapDispatchToProps = dispatch => ({
  dispatchDeleteContact: id => dispatch(deleteContact(id)),
});

export default connect(mapStateToProps, mapDispatchToProps)(SingleContactPage);
