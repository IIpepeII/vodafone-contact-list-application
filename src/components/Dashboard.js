/*
* Dashboard component. It is the initial page.
* Props:
* - filteredContacts: contacts state from store filtered with function in selectors/contacts.js
* - contacts: unfiltered contacts state from store. Needed to display letters for filter
* - filter: filter state to add active class to filter letters
* - loading: loading state to decide whether display <Loading /> or contact list
* - dispatchTextFilter: send filter text to store via setTextFilter action creator
*/
import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import filterContacts from '../selectors/contacts';
import ContactCard from './ContactCard';
import Loading from './Loading';
import setTextFilter from '../actions/filters';

export class Dashboard extends Component {
  static propTypes = {
    filteredContacts: PropTypes.arrayOf(PropTypes.object),
    contacts: PropTypes.arrayOf(PropTypes.object),
    filter: PropTypes.string,
    loading: PropTypes.bool.isRequired,
    dispatchTextFilter: PropTypes.func.isRequired,
  }

  static defaultProps = {
    filteredContacts: [],
    contacts: [],
    filter: '',
  }

  // Method for getting individual first letters from last names for filtering
  getFirstLettersFromLastNames = () => {
    const letters = this.props.contacts
      .map(contact => contact.lastName.charAt(0))
      .sort();
    return [...new Set(letters)];
  }

  render() {
    const {
      filteredContacts,
      dispatchTextFilter,
      filter,
    } = this.props;

    return (
      <div className="dashboard">
        <h1 className="text-center">Contacts</h1>
        <div className="text-center pt-3">
          {
            this.props.contacts.length > 0
            && this.getFirstLettersFromLastNames().map(letter => (
              <button
                type="button"
                className={`text-uppercase btn--filter mx-2 ${filter === letter ? 'active' : ''}`}
                id={`filter-${letter}`}
                key={letter}
                onClick={() => dispatchTextFilter(letter)}
              >
                { letter }
              </button>
            ))
          }
        </div>
        {/* Display 'Clear filter' button if contacts are filtered */}
        {
          filter.length > 0 && (
          <div className="text-center py-3">
            <button
              type="button"
              className="btn btn--empty btn--red-empty"
              onClick={() => dispatchTextFilter('')}
            >
            Clear filter
            </button>
          </div>
          )
        }
        {/* If contactsAreLoading state is true display <Loading /> else display contacts */}
        <div className="row mt-5">
          {
            this.props.loading
              ? <Loading />
              : filteredContacts.map(contact => <ContactCard data={contact} key={contact.id} />)
          }
        </div>
      </div>
    );
  }
}

/*
* Get states from the Redux store to add as props to the component
*/
const mapStateToProps = state => ({
  filteredContacts: filterContacts(state.contacts, state.filters.text),
  contacts: state.contacts,
  filter: state.filters.text,
  loading: state.contactsAreLoading,
});

/*
* Create dispatchTextFilter function to link setTextFilter action creator with the component via props
*/
const mapDispatchToProps = dispatch => ({
  dispatchTextFilter: text => dispatch(setTextFilter(text)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Dashboard);
