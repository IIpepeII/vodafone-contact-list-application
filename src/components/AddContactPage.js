/*
* Add new contact component. I use shortid package to generate id.
* The form is in ./ContactForm.js <ContactForm />
* Props:
* - dispatchAddContact for sending data to the store via addContact action creator.
* - React Router history object for dropping user back to dashboard after saving the form.
*/
import React, { Component } from 'react';
import { connect } from 'react-redux';
import shortid from 'shortid';
import PropTypes from 'prop-types';
import { addContact } from '../actions/contacts';
import ContactForm from './ContactForm';

class AddContactPage extends Component {
  static propTypes = {
    dispatchAddContact: PropTypes.func.isRequired,
    history: PropTypes.object.isRequired,
  }

  // Create contact object then sending data to the store then drop to dashboard.
  handleAddContact = (data) => {
    const contact = {
      id: shortid.generate(),
      ...data,
    };

    this.props.dispatchAddContact(contact);
    this.props.history.push('/');
  }

  render() {
    return (
      <div>
        <h1>Add New Contact</h1>
        {/* ContactForm component render text on form button based on page prop. */}
        <ContactForm
          getFormData={this.handleAddContact}
          page="addContact"
        />
      </div>
    );
  }
}

/*
* Create dispatchAddContact function to link addContact action creator with the component via props.
*/
const mapDispatchToProps = dispatch => ({
  dispatchAddContact: contact => dispatch(addContact(contact)),
});

export default connect(null, mapDispatchToProps)(AddContactPage);
