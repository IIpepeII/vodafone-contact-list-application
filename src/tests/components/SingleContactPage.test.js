import React from 'react';
import { shallow } from 'enzyme';
import { SingleContactPage } from '../../components/SingleContactPage';
import contacts from '../fixtures/contacts';

let dispatchDeleteContact, history, wrapper;

beforeEach(() => {
  dispatchDeleteContact = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(<SingleContactPage contact={contacts[0]} dispatchDeleteContact={dispatchDeleteContact} history={history} />);
});

test('should render SingleContactPage correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle dispatchDeleteContact correctly', () => {
  wrapper.find('#delete').simulate('click');
  expect(dispatchDeleteContact).toHaveBeenLastCalledWith(contacts[0].id);
  expect(history.push).toHaveBeenLastCalledWith('/');
});
