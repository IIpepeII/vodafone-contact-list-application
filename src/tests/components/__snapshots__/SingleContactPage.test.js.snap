// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render SingleContactPage correctly 1`] = `
<div
  className="single-contact"
>
  <h1>
    1first
     
    1last
    's Profile
  </h1>
  <div
    className="single-contact__data px-lg-7 pb-5"
  >
    <p>
      <span
        className="label-color"
      >
        Name:
      </span>
       
      1first
       
      1last
    </p>
    <p>
      <span
        className="label-color"
      >
        E-mail:
      </span>
       
      <a
        href="mailto:first1.last1@example.com"
      >
        first1.last1@example.com
      </a>
    </p>
    <p>
      <span
        className="label-color"
      >
        Phone:
      </span>
       
      741852
    </p>
  </div>
  <div
    className="text-center"
  >
    <button
      className="btn btn--empty btn--red-empty mr-4"
      onClick={[Function]}
      type="button"
    >
      Delete
    </button>
    <Link
      className="btn btn--empty btn--gray-empty"
      replace={false}
      to="/editcontact/111"
    >
      Edit
    </Link>
  </div>
  <Modal
    autoFocus={true}
    backdrop={true}
    backdropTransition={
      Object {
        "mountOnEnter": true,
        "timeout": 150,
      }
    }
    centered={false}
    className=""
    fade={true}
    isOpen={false}
    keyboard={true}
    modalTransition={
      Object {
        "timeout": 300,
      }
    }
    onClosed={[Function]}
    onOpened={[Function]}
    role="dialog"
    toggle={[Function]}
    zIndex={1050}
  >
    <ModalHeader
      closeAriaLabel="Close"
      tag="h5"
      toggle={[Function]}
      wrapTag="div"
    >
      Delete contact
    </ModalHeader>
    <ModalBody
      tag="div"
    >
      Are you sure you want to delete 
      1first
       
      1last
      's Profile?
    </ModalBody>
    <ModalFooter
      tag="div"
    >
      <Button
        className="btn btn--empty btn--red-empty"
        color="secondary"
        id="delete"
        onClick={[Function]}
        tag="button"
      >
        Delete
      </Button>
       
      <Button
        className="btn btn--empty btn--gray-empty"
        color="secondary"
        onClick={[Function]}
        tag="button"
      >
        Cancel
      </Button>
    </ModalFooter>
  </Modal>
</div>
`;
